lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/LineDeserializerUtils.java

```diff
@@ -0,0 +1,120 @@
+import io.trino.spi.type.DecimalType;
+import io.trino.spi.type.VarcharType;
+
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static io.trino.hive.formats.line.openxjson.JsonWriter.canonicalizeJsonString;
+import static io.trino.hive.formats.line.openxjson.JsonWriter.writeJsonArray;
+import static io.trino.hive.formats.line.openxjson.JsonWriter.writeJsonObject;
+import static io.trino.spi.StandardErrorCode.BAD_DATA;
+import static io.trino.spi.type.Chars.truncateToLengthAndTrimSpaces;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static io.trino.spi.type.Varchars.truncateToLength;
+
+public final class LineDeserializerUtils
+{
+    private LineDeserializerUtils() {}
+
+    public static RuntimeException parseError(String message)
+    {
+        return parseError(message, Optional.empty());
+    }
+
+    public static RuntimeException parseError(String message, Optional<Throwable> throwable)
+    {
+        String errorPrefix = "Failed to parse value: ";
+        return throwable.map(e -> new TrinoException(BAD_DATA, errorPrefix + message, e)).orElseGet(() -> new TrinoException(BAD_DATA, errorPrefix + message));
+    }
+
+    public static void writeDouble(BlockBuilder builder, Double value)
+    {
+        try {
+            DOUBLE.writeDouble(builder, value);
+        }
+        catch (TrinoException | IllegalArgumentException | UnsupportedOperationException e) {
+            throw parseError(e.getMessage(), Optional.ofNullable(e.getCause()));
+        }
+    }
+
+    public static void writeSlice(AbstractVariableWidthType type, BlockBuilder builder, Slice value)
+    {
+        try {
+            type.writeSlice(builder, value);
+        }
+        catch (TrinoException | IllegalArgumentException | UnsupportedOperationException e) {
+            throw parseError(e.getMessage(), Optional.ofNullable(e.getCause()));
+        }
+    }
+
+    public static void writeDecimal(DecimalType type, BlockBuilder builder, String value, Optional<BigDecimal> bigDecimal)
+    {
+        try {
+            bigDecimal.ifPresentOrElse(
+                    d -> HiveFormatUtils.writeDecimal(value, type, builder, d),
+                    () -> HiveFormatUtils.writeDecimal(value, type, builder));
+        }
+        catch (TrinoException | IllegalArgumentException | UnsupportedOperationException e) {
+            throw parseError(e.getMessage(), Optional.ofNullable(e.getCause()));
+        }
+    }
+
+    public static void writeTimestamp(TrinoTimestampEncoder<?> type, BlockBuilder builder, DecodedTimestamp value)
+    {
+        try {
+            type.write(value, builder);
+        }
+        catch (TrinoException | IllegalArgumentException | UnsupportedOperationException e) {
+            throw parseError(e.getMessage(), Optional.ofNullable(e.getCause()));
+        }
+    }
+
+    public static void writeJson(AbstractVariableWidthType type, BlockBuilder builder, Object jsonValue)
```

Comment:
Overload to specialize for `VarcharType` and `CharType` and remove the possibility of calling with some other `AbstractVariableWidthType` subclass.